src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,441 @@
-   * user code, it was easy for user code to accidentally build large chains of
-   * promises and thereby leak memory.
-   *
-   * The problem of leaks is solved by automatically breaking these chains of
-   * promises, so that promises don't refer to each other in a long chain. This
-   * allows each promise to be individually collected. The idea is to "flatten"
-   * the chain of promises, so that instead of each promise pointing to its
-   * neighbour, they instead point directly the promise at the root of the
-   * chain. This means that only the root promise is referenced, and all the
-   * other promises are available for garbage collection as soon as they're no
-   * longer referenced by user code.
-   *
-   * To make the chains flattenable, the concept of linking promises together
-   * needed to become an explicit feature of the DefaultPromise implementation,
-   * so that the implementation to navigate and rewire links as needed. The idea
-   * of linking promises is based on the [[https://github.com/twitter/util/blob/master/util-core/src/main/scala/com/twitter/util/Promise.scala
-   *  Twitter promise implementation]].
-   *
-   * In practice, flattening the chain cannot always be done perfectly. When a
-   * promise is added to the end of the chain, it scans the chain and links
-   * directly to the root promise. This prevents the chain from growing forwards
-   * But the root promise for a chain can change, causing the chain to grow
-   * backwards, and leaving all previously-linked promise pointing at a promise
-   * which is no longer the root promise.
-   *
-   * To mitigate the problem of the root promise changing, whenever a promise's
-   * methods are called, and it needs a reference to its root promise it calls
-   * the `compressedRoot()` method. This method re-scans the promise chain to
-   * get the root promise, and also compresses its links so that it links
-   * directly to whatever the current root promise is. This ensures that the
-   * chain is flattened whenever `compressedRoot()` is called. And since
-   * `compressedRoot()` is called at every possible opportunity (when getting a
-   * promise's value, when adding an onComplete handler, etc), this will happen
-   * frequently. Unfortunately, even this eager relinking doesn't absolutely
-   * guarantee that the chain will be flattened and that leaks cannot occur.
-   * However eager relinking does greatly reduce the chance that leaks will
-   * occur.
-   *
-   * Future.flatMap links DefaultPromises together by calling the `linkRootOf`
-   * method. This is the only externally visible interface to linked
-   * DefaultPromises, and `linkedRootOf` is currently only designed to be called
-   * by Future.flatMap.
-   */
-  // Left non-final to enable addition of extra fields by Java/Scala converters
-  // in scala-java8-compat.
-  class DefaultPromise[T] extends AtomicReference[AnyRef](Nil) with Promise[T] {
```

Comment:
The old implementation has a *large* comment explaining its structure and how it worked, and the new one does not have such a comment. I think having the implementation explained well is very important, both for maintainability and to help others understand what is going on.

Reply:
@NthPortal I completely agree. I'll make a draft documentation to help maintainers.